llvm/lib/SYCLPostLink/ModuleSplitter.cpp

```diff
@@ -913,306 +913,134 @@ std::string ModuleDesc::makeSymbolTable() const {
-    Rules.emplace_back(Rule::RKind::K_Callback, Callback);
+void computeFuncCategoryFromAttribute(const Function &F,
+                                      const StringRef AttrName,
+                                      SmallString<256> &Result) {
+  if (F.hasFnAttribute(AttrName)) {
+    const Attribute &Attr = F.getFnAttribute(AttrName);
+    Result += Attr.getValueAsString();
   }
 
-  // Creates a simple rule, which adds a value of a string attribute into a
-  // resulting identifier.
-  void registerSimpleStringAttributeRule(StringRef AttrName) {
-    Rules.emplace_back(Rule::RKind::K_SimpleStringAttribute, AttrName);
-  }
-
-  // Creates a simple rule, which adds a value of a string metadata into a
-  // resulting identifier.
-  void registerSimpleStringMetadataRule(StringRef MetadataName) {
-    Rules.emplace_back(Rule::RKind::K_SimpleStringMetadata, MetadataName);
-  }
-
-  // Creates a simple rule, which adds one or another value to a resulting
-  // identifier based on the presence of a metadata on a function.
-  void registerSimpleFlagAttributeRule(StringRef AttrName,
-                                       StringRef IfPresentStr,
-                                       StringRef IfAbsentStr = "") {
-    Rules.emplace_back(Rule::RKind::K_FlagAttribute,
-                       Rule::FlagRuleData{AttrName, IfPresentStr, IfAbsentStr});
-  }
-
-  // Creates a simple rule, which adds one or another value to a resulting
-  // identifier based on the presence of a metadata on a function.
-  void registerSimpleFlagMetadataRule(StringRef MetadataName,
-                                      StringRef IfPresentStr,
-                                      StringRef IfAbsentStr = "") {
-    Rules.emplace_back(
-        Rule::RKind::K_FlagMetadata,
-        Rule::FlagRuleData{MetadataName, IfPresentStr, IfAbsentStr});
-  }
-
-  // Creates a rule, which adds a list of dash-separated integers converted
-  // into strings listed in a metadata to a resulting identifier.
-  void registerListOfIntegersInMetadataRule(StringRef MetadataName) {
-    Rules.emplace_back(Rule::RKind::K_IntegersListMetadata, MetadataName);
-  }
+  Result += "-";
+}
 
-  // Creates a rule, which adds a list of sorted dash-separated integers
-  // converted into strings listed in a metadata to a resulting identifier.
-  void registerListOfIntegersInMetadataSortedRule(StringRef MetadataName) {
-    Rules.emplace_back(Rule::RKind::K_SortedIntegersListMetadata, MetadataName);
+void computeFuncCategoryFromStringMetadata(const Function &F,
+                                           const StringRef MetadataName,
+                                           SmallString<256> &Result) {
+  if (F.hasMetadata(MetadataName)) {
+    const auto *MDN = F.getMetadata(MetadataName);
+    for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+      MDString *S = cast<llvm::MDString>(MDN->getOperand(I).get());
+      Result += "-" + S->getString().str();
```

Comment:
```suggestion
      Result += "-" + S->getString();
```

No need to trigger memory allocation, `SmallString::operator+=(StringRef)` exists. Even if it would require having two separate `+=` operations (for `-` and the metadata operand), I would still prefer it that way.